File: svws-webclient/core/src/asd/validate/lehrer/ValidatorLehrerPersonalabschnittsdatenRechtsverhaeltnisGeburtsdatum.ts
import { DateManager } from '../../../asd/validate/DateManager';
import { LehrerPersonalabschnittsdaten } from '../../../asd/data/lehrer/LehrerPersonalabschnittsdaten';
import { Class } from '../../../java/lang/Class';
import { ValidatorKontext } from '../../../asd/validate/ValidatorKontext';
import { LehrerRechtsverhaeltnis } from '../../../asd/types/lehrer/LehrerRechtsverhaeltnis';
import { Schuljahresabschnitt } from '../../../asd/data/schule/Schuljahresabschnitt';
import { Validator } from '../../../asd/validate/Validator';

export class ValidatorLehrerPersonalabschnittsdatenRechtsverhaeltnisGeburtsdatum extends Validator<LehrerPersonalabschnittsdaten> {

	/**
	 * Das Geburtsdatum des Lehrers
	 */
	private readonly geburtsdatum : DateManager;


	/**
	 * Erstellt einen neuen Validator mit den übergebenen Daten und dem übergebenen Kontext
	 *
	 * @param daten          die Personalabschnittsdaten für den Validator
	 * @param geburtsdatum   das Geburtsdatum des Lehrers
	 * @param kontext        der Kontext des Validators
	 */
	public constructor(daten : LehrerPersonalabschnittsdaten, geburtsdatum : DateManager, kontext : ValidatorKontext) {
		super(daten, kontext);
		this.geburtsdatum = geburtsdatum;
	}

	protected pruefe() : boolean {
		const schuljahresabschnitt : Schuljahresabschnitt | null = this.kontext().getSchuljahresabschnittByID(this.daten().idSchuljahresabschnitt);
		if (schuljahresabschnitt === null)
			return false;
		const schuljahr : number = schuljahresabschnitt.schuljahr;
		const rv : LehrerRechtsverhaeltnis | null = LehrerRechtsverhaeltnis.getBySchluessel(this.daten().rechtsverhaeltnis);
		if (rv === null) {
			this.addFehler("Kein Wert im Feld 'rechtsverhaeltnis'.");
			return false;
		}
		let success : boolean = true;
		switch (rv) {
			case LehrerRechtsverhaeltnis.L: {
				const minJahr : number = schuljahr - ((schuljahr <= 2023) ? 65 : ((schuljahr <= 2030) ? 66 : 67));
				const maxJahr : number = schuljahr - 27;
				if (!this.geburtsdatum.istInJahren(minJahr, maxJahr)) {
					this.addFehler("Der Wert für das Geburtsjahr sollte bei Beamten/-innen auf Lebenszeit (Rechtsverhältnis = L) zwischen " + minJahr + " und " + maxJahr + " liegen. Bitte prüfen!");
					success = false;
				}
				break;
			}
			case LehrerRechtsverhaeltnis.P: {
				const minJahr : number = schuljahr - 55;
				const maxJahr : number = schuljahr - 20;
				if (!this.geburtsdatum.istInJahren(minJahr, maxJahr)) {
					this.addFehler("Der Wert für das Geburtsjahr sollte bei Beamten/-innen auf Probe (Rechtsverhältnis = P) zwischen " + minJahr + " und " + maxJahr + " liegen. Bitte prüfen!");
					success = false;
				}
				break;
			}
			case LehrerRechtsverhaeltnis.W: {
				const minJahr : number = schuljahr - 50;
				const maxJahr : number = schuljahr - 18;
				if (!this.geburtsdatum.istInJahren(minJahr, maxJahr)) {
					this.addFehler("Der Wert für das Geburtsjahr sollte bei Lehramtsanwärtern/-innen (Rechtsverhältnis = W) zwischen " + minJahr + " und " + maxJahr + " liegen. Bitte prüfen!");
					success = false;
				}
				break;
			}
			default: {
				const minJahr : number = schuljahr - 80;
				const maxJahr : number = schuljahr - 18;
				if (!this.geburtsdatum.istInJahren(minJahr, maxJahr)) {
					this.addFehler("Der Wert für das Geburtsjahr sollte bei sonstigen Rechtsverhältnissen zwischen " + minJahr + " und " + maxJahr + " liegen. Bitte prüfen!");
					success = false;
				}
				break;
			}
		}
		return success;
	}

	transpilerCanonicalName(): string {
		return 'de.svws_nrw.asd.validate.lehrer.ValidatorLehrerPersonalabschnittsdatenRechtsverhaeltnisGeburtsdatum';
	}

	isTranspiledInstanceOf(name : string): boolean {
		return ['de.svws_nrw.asd.validate.lehrer.ValidatorLehrerPersonalabschnittsdatenRechtsverhaeltnisGeburtsdatum', 'de.svws_nrw.asd.validate.Validator'].includes(name);
	}

	public static class = new Class<ValidatorLehrerPersonalabschnittsdatenRechtsverhaeltnisGeburtsdatum>('de.svws_nrw.asd.validate.lehrer.ValidatorLehrerPersonalabschnittsdatenRechtsverhaeltnisGeburtsdatum');

}

export function cast_de_svws_nrw_asd_validate_lehrer_ValidatorLehrerPersonalabschnittsdatenRechtsverhaeltnisGeburtsdatum(obj : unknown) : ValidatorLehrerPersonalabschnittsdatenRechtsverhaeltnisGeburtsdatum {
	return obj as ValidatorLehrerPersonalabschnittsdatenRechtsverhaeltnisGeburtsdatum;
}

File: svws-asd/src/main/java/de/svws_nrw/asd/validate/lehrer/ValidatorLehrerPersonalabschnittsdatenRechtsverhaeltnisGeburtsdatum.java
package de.svws_nrw.asd.validate.lehrer;

import de.svws_nrw.asd.data.lehrer.LehrerPersonalabschnittsdaten;
import de.svws_nrw.asd.data.schule.Schuljahresabschnitt;
import de.svws_nrw.asd.types.lehrer.LehrerRechtsverhaeltnis;
import de.svws_nrw.asd.validate.DateManager;
import de.svws_nrw.asd.validate.Validator;
import de.svws_nrw.asd.validate.ValidatorKontext;
import jakarta.validation.constraints.NotNull;

/**
 * Dieser Validator führt eine Statistikprüfung auf das Geburtsdatum im Kontext des Rechtsverhältnisses
 * der Abschnittsdaten eines Lehrers einer Schule aus.
 */
public final class ValidatorLehrerPersonalabschnittsdatenRechtsverhaeltnisGeburtsdatum extends Validator<LehrerPersonalabschnittsdaten> {

	/** Das Geburtsdatum des Lehrers */
	private final @NotNull DateManager geburtsdatum;

	/**
	 * Erstellt einen neuen Validator mit den übergebenen Daten und dem übergebenen Kontext
	 *
	 * @param daten          die Personalabschnittsdaten für den Validator
	 * @param geburtsdatum   das Geburtsdatum des Lehrers
	 * @param kontext        der Kontext des Validators
	 */
	public ValidatorLehrerPersonalabschnittsdatenRechtsverhaeltnisGeburtsdatum(final @NotNull LehrerPersonalabschnittsdaten daten,
			final @NotNull DateManager geburtsdatum, final @NotNull ValidatorKontext kontext) {
		super(daten, kontext);
		this.geburtsdatum = geburtsdatum;
	}


	@Override
	protected boolean pruefe() {
		// Bestimme das Schuljahr über den Schuljahresabschnitt. Treten dabei Fehler auf, so ist dieser durch einen übergeordneten Validator zu prüfen.
		final Schuljahresabschnitt schuljahresabschnitt = kontext().getSchuljahresabschnittByID(daten().idSchuljahresabschnitt);
		if (schuljahresabschnitt == null)
			return false;
		final int schuljahr = schuljahresabschnitt.schuljahr;

		// Bestimme das Rechtsverhältnis. Ist dieses nicht angegeben, so wird im Folgenden von einem sonstigen Rechtsverhältnis ausgegangen
		final LehrerRechtsverhaeltnis rv = LehrerRechtsverhaeltnis.getBySchluessel(daten().rechtsverhaeltnis);
		if (rv == null) {
			addFehler("Kein Wert im Feld 'rechtsverhaeltnis'.");
			return false;
		}

		// Prüfe das Geburtsdatum bzw. das Alter bei den folgenden Rechtsverhältnissen...
		boolean success = true;
		switch (rv) {
			case LehrerRechtsverhaeltnis.L -> { // Beamtet auf Lebenszeit
				/* Das erste akzeptierte Geburtsjahr variiert je nach Renteneintrittsalter:
				 * - Schuljahre vor 2022: vor 1958 - Rentenaltersgrenze 65 Jahre
				 * - Schuljahr 2023: Sonderfall - Definition Rentenaltersgrenze 65 Jahre
				 * - Schuljahre 2024-2029: vor 1964 - Rentenaltersgrenze 66 Jahre
				 * - Schuljahr 2030: Sonderfall - Definition Rentenaltersgrenze 66 Jahre
				 * - Schuljahre nach 2030: ab 1964 - Rentenaltersgrenze 67 Jahre
				 *  */
				final int minJahr = schuljahr - ((schuljahr <= 2023) ? 65 : ((schuljahr <= 2030) ? 66 : 67));
				final int maxJahr = schuljahr - 27;  // Das letzte akzeptierte Geburtsjahr: vor 27 Jahren
				if (!geburtsdatum.istInJahren(minJahr, maxJahr)) {
					addFehler("Der Wert für das Geburtsjahr sollte bei Beamten/-innen auf Lebenszeit (Rechtsverhältnis = L)"
							+ " zwischen " + minJahr + " und " + maxJahr + " liegen. Bitte prüfen!");
					success = false;
				}
			}
			case LehrerRechtsverhaeltnis.P -> { // Beamtet auf Probe
				final int minJahr = schuljahr - 55; // das erste akzeptierte Geburtsjahr: vor 55 Jahren
				final int maxJahr = schuljahr - 20; // das letzte akzeptierte Geburtsjahr: vor 20 Jahren
				if (!geburtsdatum.istInJahren(minJahr, maxJahr)) {
					addFehler("Der Wert für das Geburtsjahr sollte bei Beamten/-innen auf Probe (Rechtsverhältnis = P)"
							+ " zwischen " + minJahr + " und " + maxJahr + " liegen. Bitte prüfen!");
					success = false;
				}
			}
			case LehrerRechtsverhaeltnis.W -> { // Beamtet in der Lehramtsausbildung
				final int minJahr = schuljahr - 50; // das erste akzeptierte Geburtsjahr: vor 50 Jahren
				final int maxJahr = schuljahr - 18; // das letzte akzeptierte Geburtsjahr: vor 18 Jahren
				if (!geburtsdatum.istInJahren(minJahr, maxJahr)) {
					addFehler("Der Wert für das Geburtsjahr sollte bei Lehramtsanwärtern/-innen (Rechtsverhältnis = W)"
							+ " zwischen " + minJahr + " und " + maxJahr + " liegen. Bitte prüfen!");
					success = false;
				}
			}
			default -> { // Sonstiges Rechtsverhältnis
				final int minJahr = schuljahr - 80;   // das erste akzeptierte Geburtsjahr: vor 80 Jahren
				final int maxJahr = schuljahr - 18;   // das letzte akzeptierte Geburtsjahr: vor 18 Jahren
				if (!geburtsdatum.istInJahren(minJahr, maxJahr)) {
					addFehler("Der Wert für das Geburtsjahr sollte bei sonstigen Rechtsverhältnissen"
							+ " zwischen " + minJahr + " und " + maxJahr + " liegen. Bitte prüfen!");
					success = false;
				}
			}
		}
		return success;
	}

}
